compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp

```diff
@@ -0,0 +1,273 @@
+}
+
+// We assume ADD with const input is NCHW if,
+// Input shape: (N, C, H, W)
+// Output shape: (N, C, H, W)
+// 1. Const shape is (1, C, 1, 1)
+// 2. Input, Output, Const have the same C.
+bool is_NCHW_with_const(const luci::CircleAdd *node, luci::CircleNode *&pred_node,
+                        luci::CircleConst *&beta)
+{
+  auto x = dynamic_cast<luci::CircleConst *>(node->x());
+  auto y = dynamic_cast<luci::CircleConst *>(node->y());
+
+  if (x != nullptr && y == nullptr)
+  {
+    pred_node = loco::must_cast<luci::CircleNode *>(node->y());
+    beta = x;
+  }
+  else if (x == nullptr && y != nullptr)
+  {
+    pred_node = loco::must_cast<luci::CircleNode *>(node->x());
+    beta = y;
+  }
+  else
+  {
+    // Ignore if ADD does not have a constant input.
+    return false;
+  }
+
+  const auto const_rank = beta->rank();
+  if (const_rank != 4)
+    return false;
+
+  // Check the shape is (1, C, 1, 1)
+  for (uint32_t i = 0; i < const_rank; i++)
+  {
+    if (i == 1)
+      continue;
+
+    if (beta->dim(i).value() != 1)
+      return false;
+  }
+
+  const auto const_cdim = beta->dim(1);
+  const auto input_cdim = pred_node->dim(1);
+  const auto output_cdim = node->dim(1);
+
+  // Check Input, Output, Const have the same channel size
+  if (const_cdim == input_cdim && input_cdim == output_cdim)
+    return true;
+  else
+    return false;
+}
+
+class ConvertNCHWToNHWC final : public luci::CircleNodeMutableVisitor<bool>
+{
+  // Default
+  bool visit(luci::CircleNode *node)
+  {
+    throw std::runtime_error(node->name() + " is an unsupported operator.");
+  }
+
+  bool visit(luci::CircleAdd *node)
+  {
+    luci::CircleNode *pred_node = nullptr;
+    luci::CircleConst *beta = nullptr;
+
+    if (is_NCHW_with_const(node, pred_node, beta))
+    {
+      auto pre_trans = create_pre_transpose(node);
+      pre_trans->a(pred_node);
+      node->x(pre_trans);
+
+      auto nhwc_const = create_NHWC_from_NCHW(beta);
+      node->y(nhwc_const);
+    }
+    else if (beta == nullptr)
+    {
+      // Both inputs are not constant.
+      // In this case, we cannot distinguish NCHW from NHWC,
+      // so just insert Transpose Ops.
+      auto pre_trans_x = create_pre_transpose(node);
+      pre_trans_x->a(node->x());
+      node->x(pre_trans_x);
+
+      auto pre_trans_y = create_pre_transpose(node);
+      pre_trans_y->a(node->y());
+      node->y(pre_trans_y);
+    }
+    else
+    {
+      return false;
+    }
+
+    // Make loco do shape inference for this node again.
+    loco::shape_erase(node);
+
+    auto post_trans = create_post_transpose(node);
+    loco::replace(node).with(post_trans);
+
+    post_trans->a(node);
+    return true;
+  }
+};
+
+} // namespace
+
+namespace luci
+{
+
+using DataFormat = ConvertNCHWToNHWCPass::DataFormat;
+
+bool ConvertNCHWToNHWCPass::run(loco::Graph *g)
+{
+  LOGGER(l);
+  INFO(l) << "ConvertNCHWToNHWCPass Start" << std::endl;
+
+  // Collect target operators (done once)
+  if (_op_status.empty())
```

Comment:
@jinevening What if the op, that `_op_status` has, is removed or changed? Isn't it possible to derefer null ptr?

Reply:
`_op_status` is `std::map`, declared with `ConvertNCHWToNHWCPass`, so  I think it cannot be nullptr.

I found that _op_status is public by default (the pass is `struct`, not `class`). I'll make them `private` not to be modified by others.

Reply:
@jinevening Say this node is one of `_op_status` whose name is A and node A is removed by other pass. And then, back to this pass, `_op_status` still has node A that is actually `nullptr`. I might be wrong because I just skimmed this PR. Can't it be happend?

Reply:
> node A is removed by other pass

I think this will not happen because `_op_status` is a private member of `ConvertNCHWToNHWCPass`. There is no way to access `_op_status` outside of this pass (Please tell me if I'm wrong :sweat_smile:).

Reply:
@mhs4670go  I misunderstood the comment. You're right. It is not safe to maintain `_op_status` because nodes can be erased by other passes. I think I can use `loco::NodeAnnotation` to avoid such cases. I'll update the PR soon. Thanks for the review.